<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>The hidden manifold distance for functional data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">The hidden manifold distance for functional manifold data</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Paper</a>
</li>
<li>
  <a href="sim_functional_data.html">Functional manifold simulation scenarios</a>
</li>
<li>
  <a href="test_analytic_geodesic_distance.html">Verifying analytic geodesic distance derivation</a>
</li>
<li>
  <a href="README.html">About</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">The hidden manifold distance for functional data</h1>

</div>


<div id="abstract" class="section level1">
<h1><span class="header-section-number">1</span> Abstract</h1>
<p>Certain functional data of interest reside in non-linear spaces; the class of probability density functions provides one such example. Since these spaces are not vector spaces, even basic operations such as addition and subtraction require special consideration and treatment. In this work we set ourselves the task of calculating the geodesic distance between two functions living near a non-linear manifold due to noise. This setting cannot be tackled by classic manifold learning techniques which require data to live exactly on a manifold. Good estimation of the geodesic distance has implications for many downstream tasks in functional data analysis, including functional clustering, and nonparametric functional regression.</p>
</div>
<div id="introduction-susan" class="section level1">
<h1><span class="header-section-number">2</span> Introduction [Susan]</h1>
<p>The classical setup in FDA assumes we observe a sample of <span class="math inline">\(n\)</span> independant realizations <span class="math inline">\(X_1,\ldots,X_n\)</span> of a random variable <span class="math inline">\(X\)</span> that takes value in the Hilbert space <span class="math inline">\(L^2([a,b],{\mathbb{R}})\)</span>. However, it might very well happen that the function <span class="math inline">\(X\)</span> belongs to <span class="math inline">\({\mathcal{M}}\subset L^2([a,b],{\mathbb{R}})\)</span> a low-dimensional nonlinear manifold (gives examples and references). If it is indeed the case, it may no longer be appropriate to use <span class="math inline">\(L^2\)</span> distance between two curves as a measure of distance. <!-- Scenario 2 in [sim_functional_data](sim_functional_data.html) is one such example.  --> One may want to consider using some geodesic distance which takes into account the intrinsic structure of <span class="math inline">\({\mathcal{M}}\)</span>.</p>
<p>Before continuing, some diambiguities are called for. Contrast existing work on functional data where either the domain is a manifold or the range is a manifold, or both.</p>
<p>There exist many nonlinear dimension reduction methods for manifolds embedded in Euclidean space, also known as manifold learning methods, which are particularly popular in computer vision. However, the success of these methods typically require the data to be observed with a high signal-to-noise ratio. (is it true? ref?) This is a luxury that FDA cannot always afford because functional data are usually observed with noise. The noise can be such that the observed functional data no longer live on a manifold. Though powerful techniques exist for smoothing/interpolating discretely observed noisy functional data, the recovered functional versions of the data, <span class="math inline">\(\tilde X_1,\ldots,\tilde X_n\)</span>, may not lie even close to the functional manifold <span class="math inline">\({\mathcal{M}}\)</span>. This presents challenges to the application of many manifold learning techniques which assume fully observed nearly-noiseless inputs.</p>
<p>Estimating the manifold <span class="math inline">\({\mathcal{M}}\)</span> from a sample of functions observed discretely and with measurements is a challenging task (results depend on curvature of the manifold, need a lot of data). Even after recovering a functional version of the data, the chances are pretty high that <span class="math inline">\(\tilde X_1,\ldots,\tilde X_n\)</span> do not lie exactly on <span class="math inline">\({\mathcal{M}}\)</span>.</p>
<p>With this challenge in mind, we put forth a technique for the specific task of estimating pairwise geodesic distances <span class="math inline">\(\{ d_{\mathcal{M}}(X_i,X_j\}_{i&gt;j}\)</span> when we have access only to discretely-observed noisy functional observations that possibly live off the true manifold. Pairwise distances are important for many downstream tasks, e.g. classification, clustering, manifold learning (ISOMAP), nonparametric regression (kernel regression). Certainly, pairwise geodesic distances are legitimate objects of interest in themselves but our motivation is more general in nature. Specifically, we use pairwise distances as an illuminating example to advocate and promote the nonlinear perspective in FDA.</p>
<p>???insert example that shows pairwise geodesic distances give visually more interesting results than pairwise <span class="math inline">\(L_2\)</span> distance. like the Berkeley growth curves, MDS based on pairwise geodesic versus MDS based on pairwise <span class="math inline">\(L_2\)</span>. The former separates the genders better? Something like this, but hopefully a different dataset.???</p>
<p>Briefly outline the methodology: MDS plus density ridge are used in a cool way to estimate (consistently?, let’s hope..) pairwise geodesic distances.</p>
<div id="setting-susan" class="section level2">
<h2><span class="header-section-number">2.1</span> Setting [Susan]</h2>
<p>Describe preprocessing of discretely observed functional data: In the classical approach to analysing such data, we first obtain a functional version <span class="math inline">\(\tilde X_1,\ldots,\tilde X_n\)</span> of the raw data (by either smoothing or using PACE), next perform FPCA to reduce the dimension of the data, and then proceed by employing multivariate techniques.</p>
<p>What type of noise structures are allowed/considered? What type of time sampling frequency is assumed? Do not need time-aligned curves!</p>
</div>
</div>
<div id="preliminaries-susan" class="section level1">
<h1><span class="header-section-number">3</span> Preliminaries [Susan]</h1>
<p>We first rigorously define the concepts of geodesic and geodesic distance. We will also describe an important helper algorithm which we call IsoGeo.</p>
<div id="basic-differential-geometry-susan" class="section level2">
<h2><span class="header-section-number">3.1</span> Basic differential geometry [Susan]</h2>
<p>This is take from a paper on <a href="http://proceedings.mlr.press/v32/linb14.pdf">metric learning</a>. Let <span class="math inline">\(({\mathcal{M}},g)\)</span> be a <span class="math inline">\(d\)</span>-dimensional Riemannian manifold, where <span class="math inline">\(g\)</span> is a Riemannian metric tensor on <span class="math inline">\({\mathcal{M}}\)</span>. This metric tensor can be used to assign a metric on the manifold. For each point <span class="math inline">\(p\)</span> on the manifold, the Riemannian metric tensor <span class="math inline">\(g\)</span> has an inner product <span class="math inline">\(g_p\)</span> on the tangent space <span class="math inline">\(T_p {\mathcal{M}}\)</span>. The norm of a tangent vector <span class="math inline">\(v \in T_p {\mathcal{M}}\)</span> is defined as <span class="math display">\[||v|| = \sqrt{g_p(v,v)}\]</span> If <span class="math inline">\(\gamma: [a,b] \subset {\mathbb{R}}\to {\mathcal{M}}\)</span> is a smooth curve, its length is then defined as <span class="math display">\[ l(\gamma) := \int_a^b || \frac{\,d\gamma}{\,d t}(t) || \,dt \]</span> The geodesic distance between two points <span class="math inline">\(p,q\)</span> on the manifold <span class="math inline">\({\mathcal{M}}\)</span>, based on this metric tensor <span class="math inline">\(g\)</span>, is defined as <span class="math display">\[ d_g(p,q):=\inf \{l(\gamma): \gamma:[a,b] \to {\mathcal{M}}\text{ piecewise smooth}, \gamma(a) = p, \gamma(b) = q]  \}.\]</span></p>
</div>
<div id="isogeo-marie" class="section level2">
<h2><span class="header-section-number">3.2</span> IsoGeo [Marie]</h2>
<p>Isomap is a three steps procedure that takes as input a set of points <span class="math inline">\(x_1,\ldots,x_n\in {\mathbb{R}}^D\)</span> and produces an embedding of the input data in the space <span class="math inline">\({\mathbb{R}}^d\)</span> with <span class="math inline">\(d&lt;D\)</span>, that preserves pairwise geodesic distances. * The first step consists in constructing a weighted graph <span class="math inline">\(G\)</span> with nodes corresponding to the input data and for which the weight of an edge between two nodes <span class="math inline">\(x_i\)</span> and <span class="math inline">\(x_j\)</span> is equal to <span class="math inline">\(||x_i-x_j ||_{{\mathbb{R}}^D}\)</span>. The graph <span class="math inline">\(G\)</span> is constructed such that a node is connected to a fix number <span class="math inline">\(N\)</span> of neighbors or to all nodes that are at a distance smaller than a given value <span class="math inline">\(\epsilon\)</span>. * In the second step, the pairwise geodesic distances <span class="math inline">\(d_{\mathcal{M}}(x_i,x_j)\)</span> are estimated based on <span class="math inline">\(G\)</span>. The geodesic distance between two nodes is simply the length of the shortest path in <span class="math inline">\(G\)</span> between these two nodes, and can be calculated easily with Floyd-Warshall or Dijkstra’s algorithms. * The last step consists in using MDS on the matrix of pairwise geodesic distances obtained in the first step in order to obtain an embedding in <span class="math inline">\({\mathbb{R}}^d\)</span>.</p>
<p>In what follows, we call IsoGeo the procedure of estimating pairwise geodesic distances with the two first steps of Isomap. Note that Floyd or Dijkstra are methods to find the smallest path given a weigthed graph but they are not related to how to calculate that graph (which is the difficult part, depend on number of neighbors, many method out there to calculate that graph in a robust way, p-isomap is one of these).</p>
</div>
</div>
<div id="proposed-method-for-estimating-geodesic-distances-marie" class="section level1">
<h1><span class="header-section-number">4</span> Proposed method for estimating geodesic distances [Marie]</h1>
<p>Let <span class="math inline">\(X_1,\ldots,X_n\)</span> be independent realizations of a random function <span class="math inline">\(X\in{\mathcal{M}}\subset L^2([a,b],{\mathbb{R}})\)</span>, where <span class="math inline">\({\mathcal{M}}\)</span> is an unknown manifold. Suppose that each curve <span class="math inline">\(X_i\)</span> is observed with measurements errors on a grid <span class="math inline">\(T_i=(t_{i1},\ldots,t_{iK})\)</span>, i.e. we observe a sample of <span class="math inline">\(K\)</span>-dimensional vectors <span class="math inline">\(Y_1,\ldots,Y_n\)</span> with <span class="math inline">\(Y_{ij} = X_i(t_{ij}) + \epsilon_{ij}\)</span>, where the random variables <span class="math inline">\(\epsilon_{ij}\)</span> are of mean zero and uncorrelated with each other. We assume that the grids <span class="math inline">\(T_1,\ldots,T_n\)</span> are dense.</p>
<p>Our method is described by the following steps:</p>
<ol style="list-style-type: decimal">
<li>Transform each vector <span class="math inline">\(Y_i\)</span> into a function <span class="math inline">\(\tilde X_i\)</span> by spline smoothing: <span class="math display">\[ \tilde X_i = \arg\min_{f\in C^2[0,1]}\left\{\sum_{j=1}^{K}\left(f(t_{ij})-Y_{ij}\right)^2+\lambda \|\partial^2_tf\|^2_{L^2}\right\}\]</span> where <span class="math inline">\(\lambda&gt;0\)</span> is a tuning parameter controlling the smoothness of <span class="math inline">\(\tilde X_i\)</span>.</li>
<li>Obtain a <span class="math inline">\(s\)</span>-dimensional representation <span class="math inline">\(\tilde X^s_1,\ldots,\tilde X^s_n\)</span> of the functions <span class="math inline">\(\tilde X_1,\ldots,\tilde X_n\)</span> that “preserves” the pairwise <span class="math inline">\(L^2\)</span> distances by using MDS.</li>
<li>Obtain <span class="math inline">\(\tilde X^{s,\hat {\mathcal{M}}}_1,\ldots,\tilde X^{s,\hat {\mathcal{M}}}_n\)</span> a “projection” of <span class="math inline">\(\tilde X^s_1,\ldots,\tilde X^s_n\)</span> onto a ridge <span class="math inline">\(\hat {\mathcal{M}}\)</span> which is computed with a mean shift algorithm.</li>
<li>Use IsoGeo to approximate the pairwise geodesic distances <span class="math inline">\(\{d_{\hat {\mathcal{M}}}(\tilde X^{s,\hat {\mathcal{M}}}_i,\tilde X^{s,\hat {\mathcal{M}}}_j)\}_{i&gt;j}\)</span> and define the <span class="math inline">\(n\)</span> by <span class="math inline">\(n\)</span> matrix <span class="math display">\[
\hat G(i,j)=\hat G(j,i) = \left\{ \begin{array}{ll}
 d_{\hat {\mathcal{M}}}(\tilde X^{s,\hat {\mathcal{M}}}_i,\tilde X^{s,\hat {\mathcal{M}}}_j) &amp; \textrm{if $i\neq j$,}\\
 0 &amp; \textrm{otherwise.}
  \end{array} \right.
\]</span></li>
</ol>
<p>Since the ridge estimation obtained from noisy measurements of a manifold should approximate well the manifold (ref. Genevose and al. 2014), we expect the points <span class="math inline">\(\tilde X^{s,\hat {\mathcal{M}}}_1,\ldots,\tilde X^{s,\hat {\mathcal{M}}}_n\)</span> to lie close to the real manifold <span class="math inline">\({\mathcal{M}}\)</span> and then <span class="math inline">\(d_{\hat {\mathcal{M}}}\)</span> to be close to <span class="math inline">\(d_{\mathcal{M}}\)</span>. Ridge estimation suffers from the curse of dimensionality, this is why we first reduce the dimension of our data with MDS and then apply the shift-mean algorithm to estimate the ridge.</p>
<div id="selection-of-tuning-parameters-marie" class="section level2">
<h2><span class="header-section-number">4.1</span> Selection of tuning parameters [Marie]</h2>
<p>Describe heuristics.</p>
</div>
</div>
<div id="simulation-study" class="section level1">
<h1><span class="header-section-number">5</span> Simulation study</h1>
<p>We perform a simulation study to ascertain the efficacy of our method for estimating pairwise geodesic distances for discretely-observed noisy functional data.</p>
<div id="alternative-estimators-of-geodesic-distance-marie" class="section level2">
<h2><span class="header-section-number">5.1</span> Alternative estimators of geodesic distance [Marie]</h2>
<ul>
<li><strong>(Raw Data) RD</strong> Apply IsoGeo on the raw data <span class="math inline">\(Y_1,\ldots,Y_n\)</span> to obtain an estimator <span class="math inline">\(\hat G_{\textrm{RD}}\)</span>. Note that this procedure can only be used if the grid <span class="math inline">\(T_i\)</span> is the same for each <span class="math inline">\(i\)</span>.</li>
<li><strong>(Spline Smoothing) SS</strong> Transform each vector <span class="math inline">\(Y_i\)</span> into a function <span class="math inline">\(\tilde X_i\)</span> by spline smoothing. Apply IsoGeo on the vectors <span class="math inline">\(\{(\tilde X_i(t_1), \ldots, X_i(t_K)\}_{i=1}^n\)</span>, where <span class="math inline">\(t_1,\ldots,t_K\)</span> is a regular grid of <span class="math inline">\([a,b]\)</span>, to obtain an estimator <span class="math inline">\(\hat G_{\textrm{SS}}\)</span>.</li>
<li><strong>(Penalized-Isomap) pI</strong> Transform each vector <span class="math inline">\(Y_i\)</span> into a function <span class="math inline">\(\tilde X_i\)</span> by spline smoothing. Calculate the weighted graph of step I of Isomap with a penalty (describe the penalty of Chen and Muller). Apply step II of Isomap to obtain <span class="math inline">\(\hat G_{\textrm{pI}}\)</span>.</li>
<li><strong>(Random Projection) RP</strong> same method as our but change step 2 by : obtain <span class="math inline">\(s\)</span>-dimensional representation by random projection and setp 4 by obtain <span class="math inline">\(\hat G\)</span> using a ensemble method.</li>
</ul>
</div>
<div id="performance-metrics-susan" class="section level2">
<h2><span class="header-section-number">5.2</span> Performance metrics [Susan]</h2>
<p>We describe three different metrics we use to assess the quality of a pairwise geodesic distance estimator. This is the ROC curve with <span class="math inline">\(\epsilon\)</span> on the <span class="math inline">\(x\)</span>-axis and degree to which near-<span class="math inline">\(\epsilon\)</span> isometry holds, i.e. the percentage of estimated pairwise distances between <span class="math inline">\(1-\epsilon\)</span> and <span class="math inline">\(1+\epsilon\)</span> of the truth pairwise distance.</p>
</div>
<div id="simulation-scenarios-susan" class="section level2">
<h2><span class="header-section-number">5.3</span> Simulation scenarios [Susan]</h2>
<p>Acknowledge that we are sampling on a very concentrated measure to make sure functional manifold is well sampled. Cite Diaconis et. al’s “Sampling from a manifold”. How to sample properly from a functional manifold could be interesting future work.</p>
<ul>
<li>warping manifold (scenario 5 in sim_functional_data.R)</li>
<li>pdf manifold (scenario 2 in sim_functional_data.R)</li>
<li>square root pdf manifold (scenario 4 ) ### Results</li>
</ul>
<p>Submit compare_methods.sh as slurm job using sbatch.</p>
<pre class="bash"><code>sbatch compare_methods.sh</code></pre>
<p>Alternatively, we can run the local version</p>
<pre class="r"><code>R compare_methods_local.R</code></pre>
<p>Results are visualized by invoking cluster_results.R</p>
<pre class="r"><code>R cluster_results.R</code></pre>
</div>
<div id="the-cost-of-employing-proposed-method-when-manifold-is-flat-susanmarie" class="section level2">
<h2><span class="header-section-number">5.4</span> The cost of employing proposed method when manifold is flat [Susan/Marie?]</h2>
<p>Suppose we take scenario 1 where the geodesic distance coincides with the <span class="math inline">\(L_2\)</span> distance. Then what is the cost of employing our distance estimator compared to smoothing the data and doing numerical integration to find the <span class="math inline">\(L_2\)</span> distance?</p>
</div>
</div>
<div id="distance-based-functional-classification-marie" class="section level1">
<h1><span class="header-section-number">6</span> Distance-based functional classification [Marie]</h1>
<p>In this section, we explore whether our geodesic distance estimator has benefits for downstream analysis task. There are many tasks we could consider here such as distance-based nonparametric regression and distanced-based functional clustering, but we will focus on distance-based functional classification. It must be noted that while curve alignment, also known as curve registration, is necessarily performed as a preprocessing technique prior to clustering and classification, our geodesic distance estimator allows one to forsake this step.</p>
<p>For simplicity, assume the task is binary classification. Associated to each functional object <span class="math inline">\(x\)</span> is a binary <span class="math inline">\(y\)</span> indicating class membership. Consider the classifier proposed in Ferraty and Vieu (2003,2006) which is a functional version of the Nadaraya-Watson kernel estimator of class membership probabilities: <span class="math display">\[
\hat p(y = 0 | x) \frac{ \sum_{i=1}^n K[h^{-1} d(x,x_i)] 1(y_i = 0) }{ \sum_{i=1}^n K[h^{-1} d(x,x_i)] }
\]</span> We shall compare our method to using <span class="math inline">\(L_2\)</span> distance, possibly weighted, and with curve registration already accomplished. Describe alternative methods in detail.</p>
<p>The bandwidth in the classifier should be tuned individually for each method. Also we might need to tune MDS dimension <span class="math inline">\(s\)</span> since in real data, the dimension of the manifold might be much higher than encountered in the simulation scenarios where it never goes above 2.</p>
<p>Datasets used by functional classification papers</p>
<ul>
<li>Wheat, rainfall and phoneme in Aurore’s paper “Achieving near-perfect classification for functional data”</li>
<li>Berkeley growth curves in Chen et. al 2014, Biometrics.</li>
<li>Tecator and phoneme in Mahalanobis technometrics paper.</li>
<li>yeast cell cycle gene expression (can’t find this publicly) in Leng and Muller’s “Classification using functional data analysis for temporal gene expression data”</li>
</ul>
<p>Datasets used in functional manifold papers</p>
<ul>
<li>Berkeley growth, yeast cell cycle gene expression (can’t find this publicly) in Chen and Muller</li>
<li>Tecator in Lin and Yao’s contamination paper</li>
<li>Berkeley growth, gait cycle in Jean-Michel Loubes’ robust isomap paper</li>
</ul>
<!-- ## Distance-based functional clustering -->
<!-- Describe the $k$-medoids clustering method. -->
<!-- ### Simulated data -->
<!-- ###  Real data -->
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
